src/Mapper.cpp

```diff
@@ -70,6 +70,47 @@ void Mapper::processDisjointQubitLayer(
   }
 }
 
+void Mapper::processDisjoint2qBlocksLayer(
+    std::array<std::optional<std::size_t>, MAX_DEVICE_QUBITS>& lastLayer,
+    const std::optional<std::uint16_t>& control, const std::uint16_t target,
+    qc::Operation* gate) {
+  std::size_t layer = 0;
+  if (!control.has_value()) {
+    // single qubit gates are always congruent
+    layer = lastLayer.at(target).value_or(0);
+  } else {
+    if (!lastLayer.at(*control).has_value() &&
+        !lastLayer.at(target).has_value()) {
+      layer = 0;
+    } else if (!lastLayer.at(*control).has_value()) {
+      layer = *lastLayer.at(target) + 1;
+    } else if (!lastLayer.at(target).has_value()) {
+      layer = *lastLayer.at(*control) + 1;
+    } else {
+      layer = std::max(*lastLayer.at(*control), *lastLayer.at(target)) + 1;
+      for (auto& g : layers.at(layer - 1)) {
+        if ((g.control == *control && g.target == target) ||
+            (g.control == target && g.target == *control)) {
+          // if last layer contained equivalent gate, use that layer
+          layer--;
+          break;
+        }
+      }
+    }
+    lastLayer.at(*control) = layer;
+    lastLayer.at(target)   = layer;
+  }
+
+  if (layers.size() <= layer) {
+    layers.emplace_back();
+  }
+  if (control.has_value()) {
+    layers.at(layer).emplace_back(*control, target, gate);
+  } else {
+    layers.at(layer).emplace_back(-1, target, gate);
+  }
+}
```

Comment:
Just a thought: this shares quite some code with the disjoint qubits layering and only differs in a few places.
Would it be cleaner to add a flag to the existing function (something like `collect2qBlocks`) and control the behavior based on that flag?
Or would you think that makes the function too hard to understand and rather stick with the existing solution?

Reply:
Thank you for the suggestion! I now combined the two methods